File: cpp/library/test/scanner_spec.cc
//
// Created by rick on 8/19/2022.
//

#include <gtest/gtest.h>
#include <range/v3/all.hpp>
#include <range/v3/view/zip.hpp>
#include <vector>
#include "scanner.h"

using namespace lox;
using namespace ranges;

TEST(ScannerTest, ScanEmptySource) {
  Scanner scanner("");
  auto tokens = scanner.scanTokens();
  ASSERT_EQ(tokens.size(), 1);
}

TEST(ScannerTest, ScanKeywords) {
  std::vector<std::pair<std::string, TokenType>> keywords = {
      {"(", TokenType::LEFT_PAREN},
      {")", TokenType::RIGHT_PAREN},
      {"{", TokenType::LEFT_BRACE},
      {"}", TokenType::RIGHT_BRACE},
      {",", TokenType::COMMA},
      {".", TokenType::DOT},
      {"-", TokenType::MINUS},
      {"+", TokenType::PLUS},
      {";", TokenType::SEMICOLON},
      {"/", TokenType::SLASH},
      {"*", TokenType::STAR},
      {"!", TokenType::BANG},
      {"!=", TokenType::BANG_EQUAL},
      {"=", TokenType::EQUAL},
      {"==", TokenType::EQUAL_EQUAL},
      {">", TokenType::GREATER},
      {">=", TokenType::GREATER_EQUAL},
      {"<", TokenType::LESS},
      {"<=", TokenType::LESS_EQUAL},
      {"abcd", TokenType::IDENTIFIER},
      {"\"a string\"", TokenType::STRING},
      {"123.123", TokenType::NUMBER},
      {"and", TokenType::AND},
      {"class", TokenType::CLASS},
      {"else", TokenType::ELSE},
      {"false", TokenType::FALSE},
      {"fun", TokenType::FUN},
      {"for", TokenType::FOR},
      {"if", TokenType::IF},
      {"nil", TokenType::NIL},
      {"or", TokenType::OR},
      {"print", TokenType::PRINT},
      {"return", TokenType::RETURN},
      {"super", TokenType::SUPER},
      {"this", TokenType::THIS},
      {"true", TokenType::TRUE},
      {"var", TokenType::VAR},
      {"while", TokenType::WHILE}};

  auto sources = keywords | views::keys | views::join(' ') | to<std::string>();
  auto tokenTypes = keywords | views::values | to<std::vector<TokenType>>();

  auto scanner = Scanner(sources.c_str());
  auto tokens = scanner.scanTokens();
  auto&& last = tokens[tokens.size() - 1];

  ASSERT_EQ(tokens.size(), keywords.size() + 1);
  ASSERT_EQ(last.tokenType, TokenType::EndOfFile);

  for (auto&& kv : views::zip(tokens, tokenTypes)) {
    auto&& [token, tokenType] = kv;
    ASSERT_EQ(tokenType, token.tokenType);
  }
}

File: scala/library/.jvm/src/test/scala/com/github/wenjunhuang/lox/parser/ScannerSpec.scala
package com.github.wenjunhuang.lox.parser

import com.github.wenjunhuang.lox.{Scanner, Token, TokenType, Value}
import org.scalatest.flatspec.AnyFlatSpec
import org.scalatest.matchers.should.Matchers

class ScannerSpec extends AnyFlatSpec with Matchers:
  "A Scanner" should "insert eof token at the end" in {
    val source  = ""
    val scanner = new Scanner(source)
    val tokens  = scanner.scanTokens()
    tokens.size should be(1)
    tokens(0) should be(Token(TokenType.EOF, "", Value.NoValue, 1))
  }

  it should "scan number" in {
    val source  = "123 8123.123"
    val scanner = new Scanner(source)
    val tokens  = scanner.scanTokens()
    tokens.size should be(3)
    tokens(0) should be(Token(TokenType.NUMBER, "123", Value.NumericValue(123.0), 1))
    tokens(1) should be(Token(TokenType.NUMBER, "8123.123", Value.NumericValue(8123.123), 1))
  }

  it should "scan key words" in {
    val tokenSource = List(
      "and"    -> TokenType.AND,
      "class"  -> TokenType.CLASS,
      "init"   -> TokenType.INIT,
      "else"   -> TokenType.ELSE,
      "false"  -> TokenType.FALSE,
      "for"    -> TokenType.FOR,
      "fun"    -> TokenType.FUN,
      "if"     -> TokenType.IF,
      "nil"    -> TokenType.NIL,
      "or"     -> TokenType.OR,
      "print"  -> TokenType.PRINT,
      "return" -> TokenType.RETURN,
      "super"  -> TokenType.SUPER,
      "this"   -> TokenType.THIS,
      "true"   -> TokenType.TRUE,
      "var"    -> TokenType.VAR,
      "while"  -> TokenType.WHILE,
      "+"      -> TokenType.PLUS,
      "-"      -> TokenType.MINUS,
      "*"      -> TokenType.STAR,
      "/"      -> TokenType.SLASH,
      ">"      -> TokenType.GREATER,
      "<"      -> TokenType.LESS,
      ">="     -> TokenType.GREATER_EQUAL,
      "<="     -> TokenType.LESS_EQUAL,
      "="      -> TokenType.EQUAL,
      "=="     -> TokenType.EQUAL_EQUAL,
      "!"      -> TokenType.BANG,
      "!="     -> TokenType.BANG_EQUAL,
      "("      -> TokenType.LEFT_PAREN,
      ")"      -> TokenType.RIGHT_PAREN,
      ","      -> TokenType.COMMA,
      "{"      -> TokenType.LEFT_BRACE,
      "}"      -> TokenType.RIGHT_BRACE
    )

    val scanner = new Scanner(tokenSource.map(_._1).mkString(" "))
    val tokens  = scanner.scanTokens()
    tokens.size should be(tokenSource.size + 1)

    tokenSource.zipWithIndex.foreach { case ((source, tokenType), index) =>
      tokens(index) should be(Token(tokenType, source, Value.NoValue, 1))
    }
  }

  it should "scan function call" in {
    val source  = """
                   |myFunction(1, 2, 3);
                   |otherFunction(a,b);
                   |""".stripMargin
    val scanner = new Scanner(source)
    val tokens  = scanner.scanTokens()
    tokens.size should be(17)
  }

  it should "scan return statement" in {
    val source  = """
                   |return 10;
                   |""".stripMargin
    val scanner = new Scanner(source)
    val tokens  = scanner.scanTokens()
    tokens.size should be(4)
  }

end ScannerSpec
